ncm-nfs/src/main/perl/nfs.pm

```diff
@@ -2,305 +2,468 @@
     } else {
-        $self->log("No action needed for $mntpt\n");
+        $self->verbose("$FSTAB unmodified");
     }
-  }
 
-  # Force a reload of the nfs daemon. 
-  $self->log("Forcing nfs reload\n");
-  system("service nfs reload");
-  if ($?) {
-    $self->warn("Error on nfs reload\n");
-  }
+    return $result, \%old, \@old_order, \%new, \@new_order;
+}
+
 
-  return 1;
+# Do something mount(point) related (umount, mount, remount, ...)
+# C<$cmd> is the arrayref, the mountpoint is appended
+# Returns SUCCESS on success, undef on failure.
+sub do_mount
+{
+    my ($self, $cmd, $fstab) = @_;
+
+    # Make a copy of the $cmd ref (if a Readonly array is passed,
+    # pushargs to readonly will fail)
+    my $proc = CAF::Process->new([@$cmd], log => $self);
+    $proc->pushargs($fstab->{$FSTAB_MOUNTPOINT});
+    $proc->execute();
+    if ($?) {
+        $self->warn("Error (un/re)mounting ", $fstab->{$FSTAB_MOUNTPOINT},
+                    " (device ", $fstab->{$FSTAB_DEVICE}, ", cmd $proc)");
+        return;
+    } else {
+        return SUCCESS;
+    }
 }
 
-# Compares two NFS entries for equality.  If equal, then
-# 0 is returned.  If the entries differ in the devices or
-# mountpoint, then 1 is returned.  Otherwise, 2 is returned.
-sub compare_entries($$) {
 
-    my ($nref, $oref) = @_;
+# Given the ncm-nfs config, determine the new and old ncm-nfs managed mounts
+# and do the appropriate unmounting/mounting
+sub process_mounts
+{
+    my ($self, $tree) = @_;
 
-    return 1 if ($nref->{device} ne $oref->{device});
-    return 1 if ($nref->{mntpt} ne $oref->{mntpt});
-    return 2 if ($nref->{fstype} ne $oref->{fstype});
-    return 2 if ($nref->{opt} ne $oref->{opt});
-    return 2 if ($nref->{freq} != $oref->{freq});
-    return 2 if ($nref->{passno} != $oref->{passno});
+    my ($fstab_changed, $old, $old_order, $new, $new_order) = $self->fstab($tree);
 
-    # Got the end, so hashes must be equal.
-    return 0;
+    # Unmount the file systems which are going away.
+    # Unmount in the reverse order that the file systems are defined.
+    my $action_taken = 0;
+
+    foreach my $device ( reverse @$old_order ) {
+        # Unmount volume ONLY if new entry doesn't exist or the
+        # mount point has changed.
+        if (! defined($new->{$device}) ||
+            ($new->{$device}->{$FSTAB_ACTION} eq $FSTAB_ACTION_UMOUNT_MOUNT) ) {
+
+            $action_taken = 1;
+            if ($self->do_mount(\@CMD_UMOUNT_LAZY, $old->{$device})) {
+                # TODO: previous code always did his,
+                #       without recreating the mountpoint for UMOUNT_MOUNT devices
+                #       That was probably a bug
+                if (! defined($new->{$device})) {
+                # Try removing mount point, giving warning on error.
+                    my $mntpt = $old->{$device}->{$FSTAB_MOUNTPOINT};
+                    # Disable any backup
+                    if (! $self->_cleanup($mntpt, '')) {
```

Comment:
Is it worth even removing the mount point? I would leave it alone and not clean up.